chips/nrf52/src/nfct.rs

```diff
@@ -0,0 +1,668 @@
+        let bits = self.registers.rxd_amount.read(Amount::DATABITS);
+        if bits == 0 {
+            return;
+        }
+        if amount < buf.len() {
+            // bit_and with 2^bits - 1
+            buf[amount] &= (1 << bits) - 1;
+        }
+    }
+
+    fn get_state(&self) -> NfcState {
+        self.state.get()
+    }
+
+    fn get_rx_amount(&self) -> u32 {
+        let rx_amount = self.registers.rxd_amount.read(Amount::DATABYTES)
+            + match self.registers.rxd_amount.read(Amount::DATABITS) {
+                0 => 0,
+                _ => 1,
+            };
+        // If CRC bytes are counted
+        if rx_amount > 2 {
+            return rx_amount - 2;
+        }
+        rx_amount
+    }
+
+    fn handle_selected(&self) {
+        self.state.set(NfcState::Activated);
+        self.client.map(|client| client.tag_selected());
+    }
+
+    fn handle_rxend(&self) {
+        self.state.set(NfcState::Received);
+        // Return the buffer to the capsule
+        self.client.map(|client| {
+            self.rx_buffer.take().map(|rx_buffer| {
+                let returncode = if self.registers.framestatus_rx.is_set(FrameStatus::OVERRUN) {
+                    ReturnCode::ENOMEM
+                } else if self.registers.framestatus_rx.is_set(FrameStatus::CRCERROR)
+                    || self
+                        .registers
+                        .framestatus_rx
+                        .is_set(FrameStatus::PARITYSTATUS)
+                {
+                    ReturnCode::FAIL
+                } else {
+                    ReturnCode::SUCCESS
+                };
+                let rx_amount = self.get_rx_amount() as usize;
+                self.mask_rx_bits(rx_buffer);
+                client.frame_received(rx_buffer, rx_amount, returncode);
+            });
+        });
+        self.clear_errors();
+    }
+
+    fn handle_txend(&self) {
+        self.state.set(NfcState::Transmitted);
+        // Return the buffer to the capsule
+        self.client.map(|client| {
+            self.tx_buffer.take().map(|tx_buffer| {
+                let returncode = if self
+                    .registers
+                    .errorstatus
+                    .is_set(ErrorStatus::FRAMEDELAYTIMEOUT)
+                {
+                    ReturnCode::FAIL
+                } else {
+                    ReturnCode::SUCCESS
+                };
+                client.frame_transmitted(tx_buffer, returncode);
+            });
+        });
+        self.clear_errors();
+    }
+
+    fn clear_errors(&self) {
+        self.registers
+            .errorstatus
+            .set(self.registers.errorstatus.get());
+        self.registers
+            .framestatus_rx
+            .set(self.registers.framestatus_rx.get());
+    }
+
+    pub fn handle_interrupt(&self) {
+        let saved_inter = self.registers.intenset.extract();
+        self.disable_all_interrupts();
```

Comment:
In Tock, disabling interrupts while handling the bottom half is generally not necessary, as the top half interrupt handler disables the interrupt that fired (see `arch/cortex-m/src/lib.rs` line 140) and the bottom half interrupt handler does not re-enable the interrupt until the bottom half handler (i.e. this function) has executed (see `nrf52/src/chip.rs` `service_pending_interrupts()` function).

I realize that this pattern exists in several other places in Tock, but I generally think that is a result of people basing drivers off of C drivers written for other operating systems without this property.

Reply:
I disable the interrupts as I keep track at a time of which interrupt interests the sequence of operations. For example, once the NFC tag has been selected, I'm not interested in the `select` interrupt. Also when the capsule requests a `receive` or `transmit` operation it unmasks the corresponding interrupt.

That was the sequence I adapted.

Reply:
Oh I see, so the assumption is that there will never be more than one possible outstanding interrupt at a time, and disabling all interrupts is equivalent to clearing the flag for the one interrupt that arrived